### DISCUSSION TITLE
Thinking about remote testing scenarios for v3?

### DISCUSSION THREAD

**Discussion by riverar** [xunit/xunit#2220] [Question] [answered]
I see there's v3 progress being made, but I don't see any discussion/planning around testing on remote targets. Is this something you're considering for v3? For example, I'm authoring a cross-platform library and would like to run tests against Linux, macOS, and Windows simultaneously from Visual Studio on any of these platforms (though prefer Windows). Ideally, this would be easily configurable at project scope and Just Work(tm) with existing Visual Studio test surfaces (including Live Unit Testing).

**Comment by bradwilson** [comment, marked as answer]
How do you intend to launch these from Visual Studio, if you have to run them remotely? There is no gesture that I'm aware of inside Test Explorer that would allow for such a thing.

**Reply by bradwilson** [reply]
(It's probably also worth pointing out that the new out-of-process execution system is going to use TCP sockets. Right now I'm designing that around local launch and localhost-limited sockets, but there's no reason it couldn't be adapted to remote eventually, though it would obviously require some thoughts about security.)

**Reply by riverar** [reply]
TCP sounds fine. If I can fire up a daemon on a Linux machine and xunit on Windows will properly issue platform-agnostic testrun commands, then my scenario is already more/less covered! (I can workaround the localhost-only listeners.)

Just spitballing here, but perhaps in some future, there'd be a facility to tie a test project to a set of targets (local or remote). Perhaps a `xunit.json` that contains something like:

```jsonc
{
    // ...

    "targets": [
        {
            "type": "ssh",
            "platform": "linux",
            "name": "hostname" /* Matches hostname in Options > Cross Platform > Connection Manager */
        },
        {
            "type": "local"
        }
    ]
}
```

**Reply by bradwilson** [reply]
The way it's currently designed is:

- User invokes a runner (for example, by running `xunit.v3.runner.console.exe /path/to/my/tests.exe`)
- Runner opens a TCP listener socket on a random port
- Runner (locally) launches `/path/to/my/tests.exe -tcp <port>`
- Command(s) are issued from the runner to align with the user's command line request (two obvious examples: "discover tests" and "run tests")
- A QUIT command is issued from the runner to the test assembly when the requested work is complete

So there's no daemon so to speak. You have two dynamically launched executables setting up a communication channel between them.

What you'd need in this scenario that isn't currently accounted for:

- Mechanism to build and publish executable for the remote machine
- Mechanism to distribute the built bits to the remote machine
- Mechanism to remote into the machine and launch the executable, and either open up non-localhost (so it can connect back) or do further work to use the console as a tunnel

It's not clear to me which of these parts you would expect xUnit.net to do. Also, trying to go down the tunnel route would necessarily limit this to tests which are console-based. And then you have the question of the gesture in VS which doesn't exist yet which sets all of this into motion (since the runner is the initiator of the process).

You have further issues related to debugging which are going to end up being non-trivial as well. In the VS case, we will use a strategy like https://github.com/bradwilson/VSChildDebuggingPOC in order to automatically wire up child debugging (and we have no solution yet today for VS Code). Whatever gesture you imagine exists will also need to find a way to wire up not just child process debugging, but remote process debugging, if you intend to be able to use things like Test Explorer.

**Reply by bradwilson** [reply]
For what it's worth, many of these problems exist today in device runners in VSTest, but they completely own the process of building, distributing, and running the tests on the device, as well as translating the results and sending them back to VS over whatever channel is used (I'm not sure, but it's very likely to be a TCP back-channel like we're designing for v3). They are also entirely responsible for being able to debug the code on the device.

I'm 99.999% sure that if you intend for this to be a VSTest feature, then they will entirely own the process, and expect us to provide a platform-agnostic or platform-compatible VSTest plugin (like we do today for device support).

**Reply by riverar** [reply]
I think a simpler approach would be to allow for some sort of override that skips the "open local port" logic you documented above and simply connects to a given host and port (e.g. `$env:XUNIT_TARGET=1.2.3.4:11223`). Then some enterprising external developer can handle all the rest (deployment, firewalls, plumbing, etc.) and xunit will be none the wiser. Am I missing something?

(I'm looking at this from a very restrictive .net5/netcore C# library unit testing lens.)

**Reply by bradwilson** [reply]
This is really not the hard problem to be solved (passing a hostname in addition to port when launching a test executable). What's going to be listening on the other side? Who is going to launch this? Since this now becomes a TCP port open ostensibly to the world, what security mechanisms are required to be in place to ensure that connections are only issued from legitimate clients and can't cause the box to become hacked or DoS'd?

I know you think this is seems trivial on the surface, but it's *really* not.

**Reply by riverar** [reply]
I don't believe those particular problems are problems in your domain. This is dev tooling--msvsmon, windbg, and friends operate similarly and operate with an assumption that devs know a bit better.

But I digress.

Based on this discussion, it doesn't sound like my scenario fits your plans or v3 roadmap. That's okay!

I appreciate the consideration and thanks for xunit.

**Reply by bradwilson** [reply]
I still firmly believe that the solution to this problem lies entirely in the hands of VSTest, not partially. I say this because that's how it's been done historically with VSTest with devices, which is effectively a "remote" execution. No code changes were required on our side to enable that, other than plugins supporting the device platform in question; here, where the request is to support .NET Core on Linux and macOS, even that is not necessary, since our .NET Core VSTest plugin support is already operating system agnostic.

**Reply by bradwilson** [reply]
Going to mark this as answered. Please let me know if you have any further questions.
